File: Easy/0121_best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
#/bin/python3

from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minPrice = prices[0]
        maxProfit = 0

        for price in prices:
            if price < minPrice:
                minPrice = price
            elif maxProfit < price - minPrice:
                maxProfit = price - minPrice
        return maxProfit

def main():
    tests: List[List[int]] = [
        [7,1,5,3,6,4],
        [7,6,4,3,1],
    ]

    for prices in tests:
        print(f"---\nprices: {prices}")
        print(f"maxProfit: {Solution().maxProfit(prices)}")

if __name__ == "__main__":
    main()

File: Easy/0121_best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.go
package main

import (
	"fmt"
)

func maxProfit(prices []int) int {
	minPrice := prices[0]
	maxProfit := 0

	for _, price := range prices {
		if price < minPrice {
			minPrice = price
		} else if maxProfit < price - minPrice {
			maxProfit = price - minPrice
		}
	}
	return maxProfit
}

func main() {
	tests := [][]int{
		{7,1,5,3,6,4},
		{7,6,4,3,1},
	}

	for _, prices := range tests {
		fmt.Printf("---\nprices: %v\n", prices)
		fmt.Printf("maxProfit: %d\n", maxProfit(prices))
	}
}
